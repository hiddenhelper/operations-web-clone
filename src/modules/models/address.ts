import { INamedEntity } from './general';

export enum AddressFields {
  LINE_1 = 'line1',
  LINE_2 = 'line2',
  COUNTRY = 'country',
  COUNTRY_ID = 'countryId',
  CITY = 'city',
  STATE = 'stateCode',
  STATE_NAME = 'stateName',
  ZIP_CODE = 'zipCode',
  COUNTY = 'county',
  BOROUGH = 'borough',
  LATITUDE = 'latitude',
  LONGITUDE = 'longitude',
  ATTENTION_1 = 'attention1',
  ATTENTION_2 = 'attention2',
}

export enum LocationsFields {
  NAME = 'name',
}

export enum Borough {
  MANHATTAN = 'Manhattan',
  BROOKLYN = 'Brooklyn',
  QUEENS = 'Queens',
  STATEN_ISLAND = 'Staten Island',
  THE_BRONX = 'The Bronx',
}

export interface IAddress {
  [AddressFields.LINE_1]: string;
  [AddressFields.LINE_2]: string;
  [AddressFields.CITY]: string;
  [AddressFields.STATE]: string;
  [AddressFields.ZIP_CODE]: string;
  [AddressFields.COUNTRY]?: INamedEntity;
  [AddressFields.COUNTRY_ID]?: string;
  [AddressFields.COUNTY]?: string;
  [AddressFields.ATTENTION_1]?: string;
  [AddressFields.ATTENTION_2]?: string;
  [AddressFields.STATE_NAME]?: string;
  [AddressFields.BOROUGH]?: Borough;
  [AddressFields.LATITUDE]: number;
  [AddressFields.LONGITUDE]: number;
}

export interface ILocation {
  id?: string;
  [LocationsFields.NAME]?: string;
  isUsed?: boolean;
}

export const addressFieldRules = {
  required: [
    { name: AddressFields.LINE_1, required: true },
    { name: AddressFields.LINE_2, required: false },
    { name: AddressFields.COUNTRY_ID, required: true, computePositive: true },
    { name: AddressFields.STATE, required: true },
    { name: AddressFields.ZIP_CODE, required: true },
    { name: AddressFields.CITY, required: true },
    { name: AddressFields.ATTENTION_1, required: false },
    { name: AddressFields.ATTENTION_2, required: false },
  ],
  notRequired: [
    { name: AddressFields.LINE_1, required: false },
    { name: AddressFields.LINE_2, required: false },
    { name: AddressFields.COUNTRY_ID, required: false },
    { name: AddressFields.STATE, required: false },
    { name: AddressFields.ZIP_CODE, required: false },
    { name: AddressFields.CITY, required: false },
    { name: AddressFields.ATTENTION_1, required: false },
    { name: AddressFields.ATTENTION_2, required: false },
  ],
};

export const boroughOptionList = [
  {
    value: Borough.MANHATTAN,
    label: Borough.MANHATTAN,
  },
  {
    value: Borough.BROOKLYN,
    label: Borough.BROOKLYN,
  },
  {
    value: Borough.QUEENS,
    label: Borough.QUEENS,
  },
  {
    value: Borough.STATEN_ISLAND,
    label: Borough.STATEN_ISLAND,
  },
  {
    value: Borough.THE_BRONX,
    label: Borough.THE_BRONX,
  },
];

export enum ZipCodeFormatType {
  FIVE_DIGITS = 0,
  FIVE_DIGITS_PLUS_FOUR = 1,
  CANADA_FORMAT = 2,
}

export const zipCodeFormatMap: { [key: string]: { placeholder: string; mask: any[] } } = {
  [ZipCodeFormatType.FIVE_DIGITS]: {
    placeholder: '00000',
    mask: [/\d/, /\d/, /\d/, /\d/, /\d/, /\d/],
  },
  [ZipCodeFormatType.FIVE_DIGITS_PLUS_FOUR]: {
    placeholder: '00000-0000',
    mask: [/\d/, /\d/, /\d/, /\d/, /\d/, '-', /\d/, /\d/, /\d/, /\d/],
  },
  [ZipCodeFormatType.CANADA_FORMAT]: {
    placeholder: 'ANA-NAN',
    mask: [/[a-zA-Z]/, /\d/, /[a-zA-Z]/, '-', /\d/, /[a-zA-Z]/, /\d/],
  },
};

export const getFallbackAddress = () => ({
  line1: null,
  line2: null,
  city: null,
  countryId: null,
  stateCode: null,
  zipCode: null,
  county: null,
  latitude: null,
  longitude: null,
  stateName: null,
});

export const getFallbackLocation = (): ILocation => ({
  name: null,
});
