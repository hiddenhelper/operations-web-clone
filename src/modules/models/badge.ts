import { INamedEntity } from './general';
import { WorkerProjectStatus } from './project';

export enum BadgeStatus {
  ACTIVE = 0,
  REVOKED = 1,
  DEACTIVATE = 2,
  EXPIRED = 3,
  NOT_ISSUED = 4,
}

export enum BadgeLayout {
  LARGE_HEADER = 1,
  SMALL_HEADER = 2,
  WHITE_HEADER = 3,
  VISITOR = 4,
}

export enum BadgeType {
  REGULAR = 0,
  UNIVERSAL = 1,
}

export enum VisitorAvailability {
  PENDING = 0,
  AVAILABLE = 1,
  ASSIGNED = 2,
}

export enum VisitorType {
  REGULAR = 0,
  VIP = 1,
}

export enum VisitorEntity {
  CLIENT = 0,
  OTHER = 1,
}

export enum VisitorBadgeStatus {
  AVAILABLE = 0,
  ASSIGNED = 1,
}

export interface IBadgeTemplate {
  id: string;
  layout: BadgeLayout;
  color: string;
  fontColor: string;
  font: string;
  logoUrl?: string;
  templateFileUrl?: string;
}

export interface IBadge {
  id: string;
  badgeTemplate: IBadgeTemplate;
  code: string;
  status: BadgeStatus;
  firstName: string;
  lastName: string;
  emergencyContactName: string;
  emergencyContactPhone: string;
  emergencyContactRelationship: string;
  pictureUrl: string;
  companyName: string;
  hardHatNumber: string;
  tagId?: string;
  expirationDate?: string;
  badgeType?: BadgeType;
  consentFormSigned?: boolean;
  trainingsCompleted?: boolean;
  certificationsCompleted?: boolean;
  defaultExpirationDate?: string;
  workerProjectStatus?: WorkerProjectStatus;
  isSynchronizing?: boolean;
}

export enum BadgeAction {
  UNKNOWN = 0,
  ACTIVATE = 1,
  DEACTIVATE = 2,
  REVOKE = 3,
  REACTIVATE = 4,
  ADD_TO_ACS = 5,
  REMOVE_FROM_ACS = 6,
  UPDATE_EXPIRATION_DATE = 7,
  GENERATE_PDF = 8,
  INSERT_TAG_ID = 9,
}

export interface IBadgeHistory {
  id: string;
  date: string;
  badgeAction: BadgeAction;
  tagId: string;
  description: string;
  user: {
    id: string;
    firstName: string;
    lastName: string;
    email: string;
  };
  projectAccessControlSystem: {
    id: string;
    deviceName: string;
  };
  reader: {
    id: string;
    hostname: string;
  };
}

export interface IBadgeUpdateRequest {
  expirationDate?: string;
  hasExpiration?: number;
  tagId?: string;
}

export interface IBadgeVisitor {
  id: string;
  number: number;
  code: string;
  status: BadgeStatus;
  availability: VisitorAvailability;
  firstName: string;
  lastName: string;
  visitorType: VisitorType;
  validFrom: string;
  validTo: string;
  description: string;
  company?: INamedEntity;
  companyId?: string;
  entityName: string;
  visitorBadgeStatus: VisitorBadgeStatus;
  tagId?: string;
  isSynchronizing?: boolean;
}

export const badgeActionMap = {
  [BadgeAction.UNKNOWN]: () => 'Unknown',
  [BadgeAction.ACTIVATE]: () => 'Activate the badge',
  [BadgeAction.DEACTIVATE]: () => 'Deactivate the badge',
  [BadgeAction.REVOKE]: () => 'Revoke the badge',
  [BadgeAction.REACTIVATE]: () => 'Reactivate the badge',
  [BadgeAction.ADD_TO_ACS]: (name: string) => `Badge added on ACS ${name}`,
  [BadgeAction.REMOVE_FROM_ACS]: (name: string) => `Badge removed on ACS ${name}`,
  [BadgeAction.UPDATE_EXPIRATION_DATE]: () => 'Change the expiration date',
  [BadgeAction.GENERATE_PDF]: () => 'Generate PDF',
  [BadgeAction.INSERT_TAG_ID]: () => 'Insert Tag ID',
};

export const badgeOptionList = [{ title: 'Deactivate' }, { title: 'Revoke' }, { title: 'Reprint' }];

export const badgeStatusMap = {
  [BadgeStatus.ACTIVE]: 'Active',
  [BadgeStatus.REVOKED]: 'Revoked',
  [BadgeStatus.DEACTIVATE]: 'Deactivated',
  [BadgeStatus.EXPIRED]: 'Expired',
  [BadgeStatus.NOT_ISSUED]: 'Not Issued',
};

export const modalStatusMap = {
  [BadgeStatus.ACTIVE]: () => ({
    title: 'Activate Badge?',
    text: `Enter Badge Tag ID`,
    confirmLabel: 'Activate',
    loadingText: 'Activating...',
  }),
  [BadgeStatus.DEACTIVATE]: () => ({
    title: 'Deactivate Badge?',
    text: 'Please enter the reason to deactivate the Badge.',
    confirmLabel: 'Deactivate',
    loadingText: 'Deactivating...',
  }),
  [BadgeStatus.REVOKED]: () => ({
    title: 'Revoke Badge?',
    text: 'Please enter the reason to revoke the Badge.',
    confirmLabel: 'Revoke',
    loadingText: 'Revoking...',
  }),
  [BadgeStatus.EXPIRED]: () => ({}),
};

export const visitorAvailabilityMap = {
  [VisitorAvailability.AVAILABLE]: 'Available',
  [VisitorAvailability.ASSIGNED]: 'Assigned',
};

export const visitorTypeMap = {
  [VisitorType.REGULAR]: 'Regular',
  [VisitorType.VIP]: 'VIP',
};

export const visitorEntityMap = {
  [VisitorEntity.CLIENT]: 'Client',
  [VisitorEntity.OTHER]: 'Other',
};

export const entityOptionList = [
  { value: VisitorEntity.CLIENT, label: visitorEntityMap[VisitorEntity.CLIENT] },
  { value: VisitorEntity.OTHER, label: visitorEntityMap[VisitorEntity.OTHER] },
];

export enum BadgeModalAction {
  ACTIVE = 0,
  NOT_ACTIVE = 1,
}

export const getFallbackBadgeTemplate = (): IBadgeTemplate => ({
  id: null,
  layout: BadgeLayout.LARGE_HEADER,
  color: '1E86E8',
  fontColor: '444444',
  font: 'Arial',
  logoUrl: null,
  templateFileUrl: null,
});

export const getFallbackVisitorBadgeTemplate = (): IBadgeTemplate => ({
  id: null,
  layout: BadgeLayout.VISITOR,
  color: 'FFE304',
  fontColor: '444444',
  font: 'Arial',
  logoUrl: null,
  templateFileUrl: null,
});

export const getFallbackBadge = (isVisitor = false): IBadge => ({
  id: null,
  badgeTemplate: isVisitor ? getFallbackVisitorBadgeTemplate() : getFallbackBadgeTemplate(),
  status: BadgeStatus.DEACTIVATE,
  firstName: null,
  lastName: null,
  emergencyContactName: null,
  emergencyContactPhone: null,
  emergencyContactRelationship: null,
  companyName: null,
  hardHatNumber: null,
  code: null,
  pictureUrl: null,
  tagId: null,
  expirationDate: null,
  consentFormSigned: false,
  trainingsCompleted: false,
  certificationsCompleted: false,
  badgeType: BadgeType.REGULAR,
  defaultExpirationDate: null,
  workerProjectStatus: WorkerProjectStatus.PENDING,
});

export const getFallbackBadgeVisitor = (): IBadgeVisitor => ({
  id: null,
  number: null,
  code: null,
  firstName: null,
  lastName: null,
  validFrom: null,
  validTo: null,
  companyId: null,
  entityName: null,
  description: null,
  visitorType: VisitorType.REGULAR,
  status: BadgeStatus.DEACTIVATE,
  availability: VisitorAvailability.AVAILABLE,
  visitorBadgeStatus: VisitorBadgeStatus.AVAILABLE,
  tagId: null,
});
