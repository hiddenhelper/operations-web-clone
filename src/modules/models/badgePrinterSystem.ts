import { IDevice } from './device';
import { INamedEntity } from './general';

export enum PrinterType {
  ZEBRA = 0,
  DATA_CARD = 1,
  BROTHER = 2,
}

export interface IBadgePrintingDevice extends IDevice {
  price: number;
  vendor: string;
  orderDate: string;
  invoice: string;
  model: string;
  inServiceDate: string;
  warrantyExpirationDate: string;
}

export interface ILaptop extends IBadgePrintingDevice {
  osVersion: string;
}

export interface IPrinter extends IBadgePrintingDevice {
  type: PrinterType;
  lastMaintenanceDate: string;
}

export interface IProjectBadgePrintingSystem extends IDevice {
  badgePrintingSystem?: INamedEntity;
  laptopSerialNumber?: string;
  printerSerialNumber?: string;
  scannerSerialNumber?: string;
  project?: {
    id: string;
    name: string;
  };
  shippingDate?: string;
}

export interface IBadgePrintingSystem extends IProjectBadgePrintingSystem {
  name: string;
  laptop: ILaptop;
  printer: IPrinter;
  scanner: IBadgePrintingDevice;
  date?: string;
  index?: number;
}

export interface IBadgePrintingSystemUpdateDate {
  badgePrintingSystemId: string;
  shippingDate: string;
}

export const printerMap = {
  [PrinterType.ZEBRA]: 'Zebra',
  [PrinterType.DATA_CARD]: 'Data Card',
  [PrinterType.BROTHER]: 'Brother',
};

export const getFallbackPrinter = (): IPrinter => ({
  id: null,
  price: null,
  vendor: null,
  orderDate: null,
  invoice: null,
  model: null,
  inServiceDate: null,
  warrantyExpirationDate: null,
  serialNumber: null,
  notes: null,
  lastMaintenanceDate: null,
  type: null,
});

export const getFallbackLaptop = (): ILaptop => ({
  id: null,
  price: null,
  vendor: null,
  orderDate: null,
  invoice: null,
  model: null,
  inServiceDate: null,
  warrantyExpirationDate: null,
  serialNumber: null,
  notes: null,
  osVersion: null,
});

export const getFallbackScanner = (): IBadgePrintingDevice => ({
  id: null,
  price: null,
  vendor: null,
  orderDate: null,
  invoice: null,
  model: null,
  inServiceDate: null,
  warrantyExpirationDate: null,
  serialNumber: null,
  notes: null,
});

export const getFallbackBadgePrinterSystem = (): IBadgePrintingSystem => ({
  id: null,
  notes: null,
  name: null,
  shippingDate: null,
  project: null,
  laptop: getFallbackLaptop(),
  printer: getFallbackPrinter(),
  scanner: getFallbackScanner(),
});
