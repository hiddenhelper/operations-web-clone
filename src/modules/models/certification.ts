import { INamedEntity, GroupValidationType } from './general';

export interface ICertification {
  id: string;
  name: string;
}

export interface IProjectCertification extends ICertification {
  alias: string;
}

export interface ICertificationGroup {
  id: string;
  name: string;
  validationType: GroupValidationType;
  certifications: IProjectCertification[];
}

export interface IWorkerCertificationFile {
  id?: string;
  displayName: string;
  url: string;
  fileSize: number;
}

export interface IWorkerCertification {
  id?: string;
  certificationId?: string;
  idNumber: string;
  projectId?: string;
  certification?: INamedEntity;
  description: string;
  completionDate: string;
  expirationDate: string;
  project?: INamedEntity;
  files?: IWorkerCertificationFile[];
  index?: number;
}

export const getCertificationFallback = (): IWorkerCertification => ({
  id: null,
  certificationId: '',
  idNumber: null,
  projectId: '',
  description: null,
  completionDate: null,
  expirationDate: null,
});
