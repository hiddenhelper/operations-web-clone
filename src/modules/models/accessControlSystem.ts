import { ILocation } from './address';
import { IDevice, DeviceStatus } from './device';

export enum AccessControlSystemType {
  PORTAL = 0,
  TURNSTILE = 1,
  ENCLOSED_TS = 2,
  HANDHELD = 3,
}

export enum ReaderType {
  Entrance = 0,
  Exit = 1,
  BiDirectionalByAntenna = 2,
  BiDirectionalBySpec = 3,
}

export interface IReader {
  serialNumber: string;
  lastMaintenanceDate?: string;
  inServiceDate: string;
  model: string;
  notes?: string;
  hostname?: string;
  directionType?: ReaderType;
  hostAddress?: string;
  telnetConnectionPort?: string;
  sshConnectionPort?: string;
  httpConnectionPort?: string;
  tcpConnectionPort?: string;
}

export enum LifeCycleType {
  NEW = 0,
  UPGRADE = 1,
  IN_FIELD_TRANSFER = 2,
  REFURBISHED = 3,
}

export enum AccessControlSystemVersion {
  V1 = 0,
  V2 = 1,
  V3 = 2,
  V3_FACIAL = 3,
}

export interface IAccessControlSystem extends IDevice {
  type: AccessControlSystemType;
  lifeCycle: LifeCycleType;
  lastRefurbishedDate: string;
  lastMaintenanceDate: string;
  inServiceDate: string;
  warrantyExpirationDate: string;
  orderDate?: string;
  vendor?: string;
  invoice?: string;
  reader1?: IReader;
  reader2?: IReader;
  version: AccessControlSystemVersion;
  project?: {
    id: string;
    name: string;
  };
  status: DeviceStatus;
  location?: ILocation;
  hasReverseInstallation?: boolean;
  reader1SerialNumber?: string;
  reader2SerialNumber?: string;
  reader1Hostname?: string;
  reader2Hostname?: string;
}

export interface IProjectAccessControlSystem {
  id?: string;
  accessControlSystemId?: string;
  projectId?: string;
  type: AccessControlSystemType;
  serialNumber: string;
  locationId?: string;
  location?: ILocation;
  deviceName: string;
  hasReverseInstallation: boolean;
  reader1: IReader;
  reader2?: IReader;
}

export interface IProjectAccessControlSystemByLocation {
  location?: ILocation;
  accessControlSystems: IAccessControlSystem[];
}

export const accessControlSystemTypeMap = {
  [AccessControlSystemType.ENCLOSED_TS]: 'Enclosed TS',
  [AccessControlSystemType.HANDHELD]: 'Handheld',
  [AccessControlSystemType.PORTAL]: 'Open Portal',
  [AccessControlSystemType.TURNSTILE]: 'Turnstile',
};

export const typeOptionList = [
  { value: AccessControlSystemType.PORTAL, label: accessControlSystemTypeMap[AccessControlSystemType.PORTAL] },
  { value: AccessControlSystemType.TURNSTILE, label: accessControlSystemTypeMap[AccessControlSystemType.TURNSTILE] },
  { value: AccessControlSystemType.ENCLOSED_TS, label: accessControlSystemTypeMap[AccessControlSystemType.ENCLOSED_TS] },
  { value: AccessControlSystemType.HANDHELD, label: accessControlSystemTypeMap[AccessControlSystemType.HANDHELD] },
];

export const lifeCycleMap = {
  [LifeCycleType.NEW]: 'New',
  [LifeCycleType.UPGRADE]: 'Upgrade',
  [LifeCycleType.IN_FIELD_TRANSFER]: 'In Field Transfer',
  [LifeCycleType.REFURBISHED]: 'Refurbished',
};

export const lifeCycleOptionList = [
  { value: LifeCycleType.NEW, label: lifeCycleMap[LifeCycleType.NEW] },
  { value: LifeCycleType.UPGRADE, label: lifeCycleMap[LifeCycleType.UPGRADE] },
  { value: LifeCycleType.IN_FIELD_TRANSFER, label: lifeCycleMap[LifeCycleType.IN_FIELD_TRANSFER] },
  { value: LifeCycleType.REFURBISHED, label: lifeCycleMap[LifeCycleType.REFURBISHED] },
];

export const deviceVersionMap = {
  [AccessControlSystemVersion.V1]: 'V1',
  [AccessControlSystemVersion.V2]: 'V2',
  [AccessControlSystemVersion.V3]: 'V3',
  [AccessControlSystemVersion.V3_FACIAL]: 'V3Facial',
};

export const deviceVersionOptionList = [
  { value: AccessControlSystemVersion.V1, label: deviceVersionMap[AccessControlSystemVersion.V1] },
  { value: AccessControlSystemVersion.V2, label: deviceVersionMap[AccessControlSystemVersion.V2] },
  { value: AccessControlSystemVersion.V3, label: deviceVersionMap[AccessControlSystemVersion.V3] },
  { value: AccessControlSystemVersion.V3_FACIAL, label: deviceVersionMap[AccessControlSystemVersion.V3_FACIAL] },
];

export const readerTypeMap = {
  [ReaderType.Entrance]: 'Entrance',
  [ReaderType.Exit]: 'Exit',
  [ReaderType.BiDirectionalByAntenna]: 'BiDirectional By Antenna',
  [ReaderType.BiDirectionalBySpec]: 'BiDirectional By Spec',
};

export const readerTypeByTypeMap = {
  [AccessControlSystemType.PORTAL]: [
    { label: readerTypeMap[ReaderType.Entrance], value: ReaderType.Entrance },
    { label: readerTypeMap[ReaderType.Exit], value: ReaderType.Exit },
    { label: readerTypeMap[ReaderType.BiDirectionalByAntenna], value: ReaderType.BiDirectionalByAntenna },
  ],
  [AccessControlSystemType.TURNSTILE]: [{ label: readerTypeMap[ReaderType.BiDirectionalBySpec], value: ReaderType.BiDirectionalBySpec }],
  [AccessControlSystemType.ENCLOSED_TS]: [{ label: readerTypeMap[ReaderType.BiDirectionalBySpec], value: ReaderType.BiDirectionalBySpec }],
  [AccessControlSystemType.HANDHELD]: [],
};

export const getFallbackAccessControlSystem = (): IAccessControlSystem => ({
  id: null,
  type: AccessControlSystemType.PORTAL,
  serialNumber: null,
  lifeCycle: LifeCycleType.NEW,
  lastRefurbishedDate: null,
  lastMaintenanceDate: null,
  inServiceDate: null,
  warrantyExpirationDate: null,
  orderDate: null,
  vendor: null,
  invoice: null,
  project: null,
  status: null,
  notes: null,
  version: AccessControlSystemVersion.V1,
  reader1: getFallbackReader(),
  reader2: null,
  hasReverseInstallation: false,
});

export const getFallbackReader = (): IReader => ({
  serialNumber: null,
  lastMaintenanceDate: null,
  inServiceDate: null,
  model: null,
  notes: null,
});

export const getFallbackProjectAccessControlSystem = (): IProjectAccessControlSystem => ({
  type: AccessControlSystemType.PORTAL,
  accessControlSystemId: null,
  serialNumber: null,
  deviceName: null,
  hasReverseInstallation: false,
  reader1: null,
  reader2: null,
});
