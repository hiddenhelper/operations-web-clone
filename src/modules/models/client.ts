import { IStep, INamedEntity } from './general';
import { IUser, UserFields } from './user';
import { IAddress, getFallbackAddress, addressFieldRules } from './address';
import { CompanyStatus } from './resource';
import { CompanyRole, CompanyProjectStatus } from './project';

export enum ClientFields {
  NAME = 'name',
  DEVELOPER = 'isDeveloper',
  TAXPAYER_IDENTIFICATION_NUMBER = 'taxpayerIdentificationNumber',
  MWBE_TYPE = 'mwbeTypeId',
  TRADES = 'trades',
  OTHER_TRADE = 'otherTrade',
  USERS = 'users',
  ADMIN_USER = 'user',
  BILLING_ADDRESS = 'billingAddress',
  MAILING_ADDRESS = 'mailingAddress',
  SAME_ADDRESS = 'mailingAddressMatchesBillingAddress',
  UNIVERSAL_BADGE_PRICE = 'universalBadgePrice',
}

export interface IClient {
  id: string;
  [ClientFields.NAME]: string;
  [ClientFields.DEVELOPER]: boolean;
  [ClientFields.TAXPAYER_IDENTIFICATION_NUMBER]?: string;
  [ClientFields.MWBE_TYPE]?: string;
  [ClientFields.TRADES]: INamedEntity[];
  [ClientFields.OTHER_TRADE]: string;
  [ClientFields.USERS]?: IUser[];
  [ClientFields.ADMIN_USER]?: IUser;
  [ClientFields.BILLING_ADDRESS]: IAddress;
  [ClientFields.MAILING_ADDRESS]: IAddress;
  [ClientFields.SAME_ADDRESS]: boolean;
  hasUniversalBadge: boolean;
  [ClientFields.UNIVERSAL_BADGE_PRICE]: number;
  status: CompanyStatus;
  hasPaymentMethod?: string;
  isTaxExempt?: boolean;
}

export interface IClientProject extends IClient {
  companyProjectStatus: CompanyProjectStatus;
  jobSiteAddress: IAddress;
  role: CompanyRole;
}

export interface IHirearchyClientProject extends Partial<IClientProject> {
  level: number;
  sponsoredCompaniesIds: string[];
  trades: INamedEntity[];
  createdAt?: string;
}

export enum ClientStep {
  GENERAL_INFORMATION = 'general-information',
  ADDRESSES = 'addresses',
  USERS = 'users',
  REVIEW = 'review',
}

export interface IStepMap {
  [key: string]: IStep;
}

export const clientStepMap: IStepMap = {
  [ClientStep.GENERAL_INFORMATION]: {
    key: ClientStep.GENERAL_INFORMATION,
    title: 'General Information',
    subtitle: 'Create New Client',
    hideControls: false,
    order: 0,
    fields: [
      { name: ClientFields.NAME, required: true },
      { name: ClientFields.DEVELOPER, required: false },
      { name: ClientFields.TAXPAYER_IDENTIFICATION_NUMBER, required: true },
      { name: ClientFields.MWBE_TYPE, required: true, computePositive: true },
      { name: ClientFields.TRADES, required: true },
      { name: ClientFields.OTHER_TRADE, required: false },
      { name: ClientFields.UNIVERSAL_BADGE_PRICE, required: false },
    ],
  },
  [ClientStep.ADDRESSES]: {
    key: ClientStep.ADDRESSES,
    title: 'Addresses',
    subtitle: 'My Company LCC',
    hideControls: false,
    order: 1,
    fields: [
      {
        name: ClientFields.BILLING_ADDRESS,
        required: true,
        fields: addressFieldRules.required,
      },
      {
        name: ClientFields.MAILING_ADDRESS,
        required: false,
        fields: addressFieldRules.notRequired,
      },
    ],
  },
  [ClientFields.USERS]: {
    key: ClientStep.USERS,
    title: 'Users',
    subtitle: 'My Company LCC',
    hideControls: false,
    order: 2,
    fields: [
      {
        name: ClientFields.USERS,
        required: true,
        fields: [
          { name: UserFields.FIRST_NAME, required: true },
          { name: UserFields.LAST_NAME, required: true },
          { name: UserFields.EMAIL, required: false },
          { name: UserFields.TITLE, required: false },
          { name: UserFields.MOBILE_PHONE_NUMBER, required: false },
          { name: UserFields.OFFICE_PHONE_NUMBER, required: false },
          { name: UserFields.PREFERRED_CONTACT_METHOD, required: true },
          { name: UserFields.GROUP_IDS, required: false },
        ],
      },
    ],
  },
  [ClientStep.REVIEW]: {
    key: ClientStep.REVIEW,
    title: 'Review',
    subtitle: 'My Company LCC',
    hideControls: true,
    order: 3,
    fields: [],
  },
};

export const developerStepMap = {
  [ClientStep.GENERAL_INFORMATION]: {
    ...clientStepMap[ClientStep.GENERAL_INFORMATION],
    fields: [{ name: ClientFields.NAME, required: true }],
  },
  [ClientStep.ADDRESSES]: {
    ...clientStepMap[ClientStep.ADDRESSES],
    fields: [],
  },
  [ClientFields.USERS]: {
    ...clientStepMap[ClientStep.USERS],
    fields: [],
  },
  [ClientStep.REVIEW]: {
    ...clientStepMap[ClientStep.REVIEW],
    fields: [],
  },
};

export const tabList: { id: string; key: string; title: string }[] = [
  {
    id: 'projects',
    key: 'projects',
    title: 'Projects',
  },
  {
    id: 'users',
    key: 'users',
    title: 'Users',
  },
  {
    id: 'workers',
    key: 'workers',
    title: 'Workers',
  },
  {
    id: 'invoices',
    key: 'invoices',
    title: 'Invoices',
  },
  {
    id: 'information',
    key: 'information',
    title: 'Information',
  },
];

export const taxExemptOptionList = [
  { value: 1, label: 'Exempt' },
  { value: 0, label: 'Not Exempt' },
];

export const getFallbackClient = (): IClient => ({
  id: null,
  name: '',
  isDeveloper: false,
  taxpayerIdentificationNumber: null,
  mwbeTypeId: '',
  trades: [],
  otherTrade: null,
  users: [],
  billingAddress: getFallbackAddress(),
  mailingAddress: getFallbackAddress(),
  mailingAddressMatchesBillingAddress: true,
  status: CompanyStatus.DRAFT,
  hasUniversalBadge: false,
  universalBadgePrice: null,
});
