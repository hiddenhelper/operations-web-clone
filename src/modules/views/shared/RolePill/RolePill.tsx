import React, { memo, useMemo } from 'react';
import { ProjectModel } from '../../../models';

export interface IRolePillProps {
  role: ProjectModel.CompanyRole;
}

const RolePill = ({ role }: IRolePillProps) => {
  const textMap = useMemo(
    () => ({
      [ProjectModel.CompanyRole.DEVELOPER]: { x: '24', y: '40', fontSize: '11' },
      [ProjectModel.CompanyRole.GENERAL_CONTRACTOR]: { x: '26', y: '41', fontSize: '13' },
      [ProjectModel.CompanyRole.SUB_CONTRACTOR]: { x: '24', y: '40', fontSize: '11' },
      [ProjectModel.CompanyRole.DEVELOPER_PROJECT_OWNER]: { x: '26', y: '41', fontSize: '13' },
    }),
    []
  );

  return (
    <svg height="70" width="70">
      <circle cx="36" cy="36" r="18" fill="#E5E5E5" />
      <text x={textMap[role].x} y={textMap[role].y} fontFamily="Verdana" fontWeight="bold" fontSize={textMap[role].fontSize} fill="#AAAAAA">
        {ProjectModel.roleAcronymMap[role]}
      </text>
    </svg>
  );
};

export default memo(RolePill);
