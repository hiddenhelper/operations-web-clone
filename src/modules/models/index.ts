import * as GeneralModel from './general';
import * as UserModel from './user';
import * as ClientModel from './client';
import * as ProjectModel from './project';
import * as ResourceModel from './resource';
import * as WorkerModel from './worker';
import * as AccessControlSystemModel from './accessControlSystem';
import * as BadgePrintingSystemModel from './badgePrinterSystem';
import * as DeviceModel from './device';
import * as InventoryModel from './inventory';
import * as BadgeModel from './badge';
import * as CertificationModel from './certification';
import * as FileModel from './file';
import * as AddressModel from './address';
import * as ConsentFormModel from './consentForm';
import * as StatisticsModel from './statistics';
import * as TrainingModel from './training';
import * as PaymentModel from './payment';
import * as InvoiceModel from './invoice';
import * as ProcoreModel from './procore';
import * as SearchModel from './search';
import * as ProjectNewModel from './project-new';

export {
  GeneralModel,
  UserModel,
  ClientModel,
  ProjectModel,
  ResourceModel,
  AccessControlSystemModel,
  WorkerModel,
  BadgePrintingSystemModel,
  DeviceModel,
  InventoryModel,
  BadgeModel,
  CertificationModel,
  FileModel,
  AddressModel,
  ConsentFormModel,
  StatisticsModel,
  TrainingModel,
  PaymentModel,
  InvoiceModel,
  ProcoreModel,
  SearchModel,
  ProjectNewModel,
};
