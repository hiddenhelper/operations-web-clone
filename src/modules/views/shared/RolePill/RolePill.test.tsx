import React from 'react';
import { render } from '@testing-library/react';

import { ProjectModel } from '../../../models';
import RolePill, { IRolePillProps } from './RolePill';

describe('RolePill', () => {
  let props: IRolePillProps;

  beforeEach(() => {
    props = {
      role: ProjectModel.CompanyRole.DEVELOPER,
    };
  });

  it('should render DEV', () => {
    const { container } = render(<RolePill {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('should render GC', () => {
    props.role = ProjectModel.CompanyRole.GENERAL_CONTRACTOR;
    const { container } = render(<RolePill {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('should render SUB', () => {
    props.role = ProjectModel.CompanyRole.SUB_CONTRACTOR;
    const { container } = render(<RolePill {...props} />);
    expect(container).toMatchSnapshot();
  });
});
