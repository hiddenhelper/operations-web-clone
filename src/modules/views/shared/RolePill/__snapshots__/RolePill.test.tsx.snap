// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RolePill should render DEV 1`] = `
<div>
  <svg
    height="70"
    width="70"
  >
    <circle
      cx="36"
      cy="36"
      fill="#E5E5E5"
      r="18"
    />
    <text
      fill="#AAAAAA"
      font-family="Verdana"
      font-size="11"
      font-weight="bold"
      x="24"
      y="40"
    >
      DEV
    </text>
  </svg>
</div>
`;

exports[`RolePill should render GC 1`] = `
<div>
  <svg
    height="70"
    width="70"
  >
    <circle
      cx="36"
      cy="36"
      fill="#E5E5E5"
      r="18"
    />
    <text
      fill="#AAAAAA"
      font-family="Verdana"
      font-size="13"
      font-weight="bold"
      x="26"
      y="41"
    >
      GC
    </text>
  </svg>
</div>
`;

exports[`RolePill should render SUB 1`] = `
<div>
  <svg
    height="70"
    width="70"
  >
    <circle
      cx="36"
      cy="36"
      fill="#E5E5E5"
      r="18"
    />
    <text
      fill="#AAAAAA"
      font-family="Verdana"
      font-size="11"
      font-weight="bold"
      x="24"
      y="40"
    >
      SUB
    </text>
  </svg>
</div>
`;
